### ISSUE TITLE
OCR doesn't work with pics averaged on channel axis

### ISSUE THREAD

**Issue by JbFoncin** [JaidedAI/EasyOCR#189]
Hi,

Thank you very much for your work. I tried to feed your OCR with my text boxes today and I encountered a very strange result :

![very_strange](https://user-images.githubusercontent.com/17411352/88675667-73fbdf80-d0eb-11ea-8f2b-c0e01223f99d.JPG)

Averaging images on channel axis makes the model blind. You need to use cvtcolor from CV2.
The difference between two images is very subtle, (pixel variation between -3 and 3 at maximum on uint8), but it looks like it's enough to stun the model.

**Comment by rkcosmos** [comment]
Interesting, can you share what is you `read_boxes` function?

**Comment by JbFoncin** [comment]
of course : 
```
class BoxReader(easyocr.Reader):

    def read_boxes(self, image_list, decoder = 'greedy', beamWidth= 5, batch_size = 1,\
                   workers = 0, allowlist = None, blocklist = None, detail = 1,\
                   contrast_ths = 0.1,adjust_contrast = 0.5, filter_ths = 0.003,\
                   text_threshold = 0.7, low_text = 0.4, link_threshold = 0.4,\
                   canvas_size = 2560, mag_ratio = 1.,\
                   slope_ths = 0.1, ycenter_ths = 0.5, height_ths = 0.5,\
                   width_ths = 0.5, add_margin = 0.1):
        '''
        Parameters:
        file: file path or numpy-array or a byte stream object
        '''

        max_width = max([image.shape[1] for image in image_list])

        if allowlist:
            ignore_char = ''.join(set(self.character)-set(allowlist))
        elif blocklist:
            ignore_char = ''.join(set(blocklist))
        else:
            ignore_char = ''.join(set(self.character)-set(self.lang_char))

        if self.model_lang in ['chinese_tra','chinese_sim', 'japanese', 'korean']: decoder = 'greedy'
        result = get_text(self.character, imgH, int(max_width), self.recognizer, self.converter, image_list,\
                      ignore_char, decoder, beamWidth, batch_size, contrast_ths, adjust_contrast, filter_ths,\
                      workers, self.device)


        if detail == 0:
            return [item[1] for item in result]
        else:
            return result
```
I had also to remove the use of coordinates in get_text:

```
def get_text(character, imgH, imgW, recognizer, converter, img_list,\
             ignore_char = '',decoder = 'greedy', beamWidth =5, batch_size=1, contrast_ths=0.1,\
             adjust_contrast=0.5, filter_ths = 0.003, workers = 1, device = 'cpu'):
    batch_max_length = int(imgW/10)

    char_group_idx = {}
    ignore_idx = []
    for char in ignore_char:
        try: ignore_idx.append(character.index(char)+1)
        except: pass

    AlignCollate_normal = AlignCollate(imgH=imgH, imgW=imgW, keep_ratio_with_pad=True)
    test_data = ListDataset(img_list)
    test_loader = torch.utils.data.DataLoader(
        test_data, batch_size=batch_size, shuffle=False,
        num_workers=int(workers), collate_fn=AlignCollate_normal, pin_memory=True)

    # predict first round
    result1 = recognizer_predict(recognizer, converter, test_loader,batch_max_length,\
                                 ignore_idx, char_group_idx, decoder, beamWidth, device = device)

    # predict second round
    low_confident_idx = [i for i,item in enumerate(result1) if (item[1] < contrast_ths)]
    if len(low_confident_idx) > 0:
        img_list2 = [img_list[i] for i in low_confident_idx]
        AlignCollate_contrast = AlignCollate(imgH=imgH, imgW=imgW, keep_ratio_with_pad=True, adjust_contrast=adjust_contrast)
        test_data = ListDataset(img_list2)
        test_loader = torch.utils.data.DataLoader(
                        test_data, batch_size=batch_size, shuffle=False,
                        num_workers=int(workers), collate_fn=AlignCollate_contrast, pin_memory=True)
        result2 = recognizer_predict(recognizer, converter, test_loader, batch_max_length,\
                                     ignore_idx, char_group_idx, decoder, beamWidth, device = device)

    result = []
    for i, pred1 in enumerate(result1):
        if i in low_confident_idx:
            pred2 = result2[low_confident_idx.index(i)]
            if pred1[1]>pred2[1]:
                result.append( (pred1[0], pred1[1]) )
            else:
                result.append( (pred2[0], pred2[1]) )
        else:
            result.append( (pred1[0], pred1[1]) )

    #confidence_score = pred_max_prob.cumprod(dim=0)[-1]
    #if confidence_score.item() > filter_ths:
    #    print(pred, confidence_score.item())
    #else:
    #    print('not sure', pred, confidence_score.item())

    return result
```

**Comment by rkcosmos** [comment]
get_text function is designed to accept color image as input, in your first case it's greyscale. That's why model was confused. If input format is right, small variation should not matter.

**Comment by JbFoncin** [comment]
My point was not about the best way to give the model pictures but resilience. If we feed the reader with an image averaged on channel axis we would get bad results, it would be nice to make it more robust. For benchmark, I compared tesseract with your OCR, here are my results, metric is CER :

| field         |   ocr_tesseract_cer |   ocr_easyocr_cer |
|:--------------|--------------------------:|------------------------:|
| code          |                0.0529173  |                0.50297  |
| document_date |                0.0706561  |                0.165128 |
| end_date      |                0.0345871  |                0.33593  |
| name          |                0.00313896 |                0.229526 |
| regime        |                0.00893977 |                0.502239 |
| start_date    |                0.0336831  |               0.326221 |
| text_dni      |                0.0456047  |               0.220077 |
| work          |                0.0216972  |                0.416899 |

each field relies on the kind of information we try to extract (code is three uppercase letters, DNI is a letter followed par 8 numbers, other are dates or common text). I would be happy to help you improve the robustness of your model, even if it performed poorly on this set of data it was very good on other documents I tested. I would be happy to help you with it.

**Comment by rkcosmos** [comment]
I see, I agree we should make grey scale image acceptable. It's actually not just bad result. In current state it should even report error because it's unexpected format. I will add a logic to check image format and do the conversion automatically.

**Comment by rkcosmos** [comment]
with lastest update, you can now use grayscale as an input.
